pkg/reader/rmn_home_poller.go

```diff
@@ -0,0 +1,386 @@
+
+	instance, ok := getInstance(key)
+	if ok {
+		lggr.Infow("RMNHomePoller already exists, reusing instance",
+			"chainSelector", rmnHomeChainSelector,
+			"address", hexEncodedAddr,
+		)
+		return instance, nil
+	}
+
+	rmnHomeBoundContract := types.BoundContract{
+		Address: hexEncodedAddr,
+		Name:    consts.ContractNameRMNHome,
+	}
+
+	if err := contractReader.Bind(ctx, []types.BoundContract{rmnHomeBoundContract}); err != nil {
+		return nil, fmt.Errorf("failed to bind RMNHome contract: %w", err)
+	}
+
+	rmnHomeReader := newRMNHomePoller(
+		logutil.WithComponent(lggr, "RMNHomePoller"),
+		contractReader,
+		rmnHomeBoundContract,
+		pollingInterval,
+	)
+
+	instances[key] = rmnHomeReader
+	return rmnHomeReader, nil
+}
+
+func getInstance(key string) (*rmnHomePoller, bool) {
+	instance, ok := instances[key]
+	if !ok {
+		return nil, false
+	}
+
+	if notStopped := instance.sync.IfNotStopped(func() {}); !notStopped {
+		return nil, false
+	}
+
+	return instance, ok
+}
+
+// newRMNHomePoller creates a new rmnHomePoller instance. Only meant to be used by getRMNHomePoller
+// or in tests. Never use that directly in the production code otherwise it will create to exhaustive polling.
+func newRMNHomePoller(
+	lggr logger.Logger,
+	contractReader contractreader.ContractReaderFacade,
+	rmnHomeBoundContract types.BoundContract,
+	pollingInterval time.Duration,
+) *rmnHomePoller {
+	return &rmnHomePoller{
+		stopCh:               make(chan struct{}),
+		contractReader:       contractReader,
+		rmnHomeBoundContract: rmnHomeBoundContract,
+		rmnHomeState:         rmnHomeState{},
+		observersMu:          &sync.Mutex{},
+		rmnHomeStateMu:       &sync.RWMutex{},
+		failedPolls:          atomic.Uint32{},
+		lggr:                 lggr,
+		pollingDuration:      pollingInterval,
+	}
+}
+
+func (r *rmnHomePoller) Start(_ context.Context, caller RMNHome) error {
+	r.observersMu.Lock()
+	defer r.observersMu.Unlock()
+
+	r.observers = append(r.observers, caller)
+
+	if len(r.observers) > 1 {
+		return nil
+	}
+
+	return r.sync.StartOnce(r.Name(), func() error {
+		r.lggr.Infow("Start Polling RMNHome")
+		r.wg.Add(1)
+		go r.poll()
+		return nil
+	})
+}
+
+func (r *rmnHomePoller) Close(caller RMNHome) error {
+	r.observersMu.Lock()
+	defer r.observersMu.Unlock()
+
+	r.observers = slices.DeleteFunc(r.observers, func(r RMNHome) bool {
+		return r == caller
+	})
+
+	if len(r.observers) > 0 {
+		return nil
+	}
+
+	err := r.sync.StopOnce(r.Name(), func() error {
+		defer r.wg.Wait()
+		close(r.stopCh)
+		return nil
+	})
+
+	if errors.Is(err, services.ErrAlreadyStopped) {
+		return nil
+	}
+
+	return err
+}
+
+func (r *rmnHomePoller) Ready() error {
+	return r.sync.Ready()
+}
+
+func (r *rmnHomePoller) HealthReport() map[string]error {
+	f := r.failedPolls.Load()
+	if f >= maxFailedPolls {
+		err := fmt.Errorf("polling failed %d times in a row (maximum allowed: %d)", f, maxFailedPolls)
+		r.sync.SvcErrBuffer.Append(err)
```

Comment:
I wonder if this err buffer needs to be cleared before appending? Could this be a potential memory leak situation?

Reply:
Actually no need, the `Healthy` call below will call `Flush` on the buffer and clear it